lib/global-styles.php

```diff
@@ -49,15 +49,43 @@ function gutenberg_experimental_global_styles_get_stylesheet( $tree, $type = 'al
  */
 function gutenberg_experimental_global_styles_enqueue_assets() {
 	if (
-		! get_theme_support( 'experimental-link-color' ) && // link color support needs the presets CSS variables regardless of the presence of theme.json file.
-		! WP_Theme_JSON_Resolver_Gutenberg::theme_has_support() ) {
+		! WP_Theme_JSON_Resolver_Gutenberg::theme_has_support() &&
+		! current_theme_supports( 'experimental-link-color' ) &&
+		current_theme_supports( 'editor-color-palette' ) &&
+		current_theme_supports( 'editor-gradient-presets' )
+	) {
 		return;
 	}
 
+	/*
+	 * We need to enqueue different styles for different situations:
+	 *
+	 * - Themes with theme.json support => everything.
+	 * - Themes without theme.json support but legacy link support => the color variables.
+	 * - Themes without theme.json support and no own color palette => the color classes + related variables.
+	 * - Themes without theme.json support and no own gradient palette => the gradient classes + related variables.
```

Comment:
This logic seems misplaced to me. It feels like instead of doing all this adhoc check, we should just build the right "theme.json" for classic themes depending on their theme support flags (logic I thought we already had) and in the end just get the full stylesheet no matter how the theme config is provided.

Reply:
I'm going to look into absorbing this logic in another place.

Note that this is the first time we're trying to do something like this: enqueue only the parts of the theme.json file that a theme uses (classic themes, in this case), hence why we need something akin to the `filter_by` function https://github.com/WordPress/gutenberg/pull/34067#discussion_r688394494 So far, we've always enqueued the whole stylesheet.